### PR TITLE
Add oracles examples

### LINKED ISSUE #25: Finish Oracle Smart Contracrts
After the recent changes, add missing unit tests for the contracts and push them to the examples repo.

child of #2

### PR DESCRIPTION
**Pull request by maptuhec**
Add oracles examples for exchange market.

closes #25 

PS: Everyone can add more reviewers

### CODE REVIEW

File: examples/ExchangeOracles/contracts/ExchangeMarket.aes [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+      
+  function updateAePrice(_aePrice: int) =
+    require(_aePrice > 0, "The price must be greater than zero")
+    put(state{aePrice = _aePrice})
+
+  function updateTokenPrice(_tokenPrice: int) =
+    require(_tokenPrice > 0, "The price must be greater than zero")
+    put(state{tokenPrice = _tokenPrice})
+
+  private function require(b : bool, err : string) =
```
Comment (alekseytsekov) [inline examples/ExchangeOracles/contracts/ExchangeMarket.aes]: b can be 'expression'

File: examples/ExchangeOracles/contracts/ExchangeMarket.aes [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+  public function getAePrice() : int = state.aePrice
+  public function getTokenPrice() : int = state.tokenPrice
+
+  function createQuery(oracle: oracle_id, question: query_type, qfee: int, qttl: int, rttl: int) : query_id =
+    require(qfee > 0 , "Query fee must be greater than zero")
+    require(qttl > 0 , "Query TTL must be greater than zero")
+    require(rttl > 0 , "Relative TTL must be greater than zero")
+    Oracle.query(oracle, question, qfee, RelativeTTL(qttl), RelativeTTL(rttl))
+
+  function createAePriceQuery(oracle: oracle_id, qfee: int, qttl: int, rttl: int) : query_id =
```
Comment (alekseytsekov) [inline examples/ExchangeOracles/contracts/ExchangeMarket.aes]: maybe 'qfee' should be qFee.. in camel case variant. same for 'qttl' and 'rttl'

File: examples/ExchangeOracles/contracts/ExchangeOracle.aes [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+      oracleFee = qfee}
+
+  function registerOracle(qfee : int, ttl  : int) : oracleId =
+    require(qfee > 0, "The fee is too low")
+    require(ttl > 0, "The ttl must be greater than zero")
+    Oracle.register(Contract.address, qfee, RelativeTTL(ttl))
+
+  public function getOracle() : oracleId = state.exchangeOracle
+
+  function respondToQuestion(_queryId : queryId, r : int)  =
```
Comment (alekseytsekov) [inline examples/ExchangeOracles/contracts/ExchangeOracle.aes]: 'r' is ?

File: examples/ExchangeOracles/contracts/ExchangeMarket.aes (hunk 2 above)
Comment (mradkov) [reply examples/ExchangeOracles/contracts/ExchangeMarket.aes]: >  in camel case variant.

I think in the functional languages like Sophia snake_case is the preferred one (I'm trying to get used to it as well :D).

Comment (thepiwo) [reply examples/ExchangeOracles/contracts/ExchangeMarket.aes]: I agree with @mradkov

File: examples/ExchangeOracles/contracts/ExchangeMarket.aes [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+  type query_type = string
+  type answer_type = int
+  type oracle_id = oracle(query_type, answer_type)
+  type query_id = oracle_query(query_type, answer_type)
+
+  public stateful function init(_aePrice: int, _tokenPrice: int) = 
+    require(_aePrice > 0 , "The price for the AE must be greater than zero ")
+    require(_tokenPrice > 0 , "The price for the token must be greater than zero")
+    {aePrice = _aePrice,
```
Comment (thepiwo) [inline examples/ExchangeOracles/contracts/ExchangeMarket.aes]: could be in one line

File: examples/ExchangeOracles/contracts/ExchangeMarket.aes [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+    require(qttl > 0 , "Query TTL must be greater than zero")
+    require(rttl > 0 , "Relative TTL must be greater than zero")
+    Oracle.query(oracle, "Give me the Token price in USD", qfee, RelativeTTL(qttl), RelativeTTL(rttl))
+
+  function queryFee(oracle : oracle_id ) : int =
+    Oracle.query_fee(oracle)
+
+  public function getAnswer(oracle: oracle_id, query : query_id) : int =
+    switch(Oracle.get_answer(oracle, query))
+            None    => 0
```
Comment (thepiwo) [inline examples/ExchangeOracles/contracts/ExchangeMarket.aes]: why not 2 spaces for indentation?

Comment (thepiwo) [reply examples/ExchangeOracles/contracts/ExchangeMarket.aes]: also why multiple spaces after none? either way line up with the other => or just one space

File: examples/ExchangeOracles/contracts/ExchangeMarket.aes (hunk 1 above)
Comment (thepiwo) [reply examples/ExchangeOracles/contracts/ExchangeMarket.aes]: @alekseytsekov why could it be?

File: examples/ExchangeOracles/contracts/ExchangeMarket.aes [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+  record state = {
+    aePrice : int,
+    tokenPrice: int}
+
+  type query_type = string
+  type answer_type = int
+  type oracle_id = oracle(query_type, answer_type)
+  type query_id = oracle_query(query_type, answer_type)
+
+  public stateful function init(_aePrice: int, _tokenPrice: int) =
```
Comment (thepiwo) [inline examples/ExchangeOracles/contracts/ExchangeMarket.aes]: sophia style usually doesn't use the _ but rather suffix `'` if it is exactly the same variable as used later as none _

**Comment by maptuhec** [comment]
> Are all the `require(... > 0` really needed? I feel most of them are overhead as the oracle system call would fail and abort anyways

In my opinion, it would be better if we have validation before the oracles/contracts revert the transaction. Also, this brings better user experience because if the user makes some little mistake with the params, the error he will get will be more descriptional.

**Comment by maptuhec** [comment]
> [aepp-sophia-examples/examples/ExchangeOracles/contracts/ExchangeMarket.aes](https://github.com/aeternity/aepp-sophia-examples/blob/0793a9f6e6f9b7b7d8e2fefc31aeda45f8ebb2d1/examples/ExchangeOracles/contracts/ExchangeMarket.aes#L20-L36)
> 
> Lines 20 to 36 in [0793a9f](/aeternity/aepp-sophia-examples/commit/0793a9f6e6f9b7b7d8e2fefc31aeda45f8ebb2d1)
> 
>  function create_query(oracle: oracle_id, question: query_type, query_fee: int, query_ttl: int, relative_ttl: int) : query_id = 
>    require(query_fee > 0 , "Query fee must be greater than zero") 
>    require(query_ttl > 0 , "Query TTL must be greater than zero") 
>    require(relative_ttl > 0 , "Relative TTL must be greater than zero") 
>    Oracle.query(oracle, question, query_fee, RelativeTTL(query_ttl), RelativeTTL(relative_ttl)) 
>   
>  function create_ae_price_query(oracle: oracle_id, query_fee: int, query_ttl: int, relative_ttl: int) : query_id = 
>    require(query_fee > 0 , "Query fee must be greater than zero") 
>    require(query_ttl > 0 , "Query TTL must be greater than zero") 
>    require(relative_ttl > 0 , "Relative TTL must be greater than zero") 
>    Oracle.query(oracle, "Give me the AE price in USD", query_fee, RelativeTTL(query_ttl), RelativeTTL(relative_ttl)) 
>   
>  function create_token_price_query(oracle: oracle_id, query_fee: int, query_ttl: int, relative_ttl: int) : query_id = 
>    require(query_fee > 0 , "Query fee must be greater than zero") 
>    require(query_ttl > 0 , "Query TTL must be greater than zero") 
>    require(relative_ttl > 0 , "Relative TTL must be greater than zero") 
>    Oracle.query(oracle, "Give me the Token price in USD", query_fee, RelativeTTL(query_ttl), RelativeTTL(relative_ttl)) 
> Why the need of 3 separate functions for the same purpose with the single difference of the question string, can't that be abstracted by using only the `create_query` fn?
> 
> Also, think this can be replaced with a single `update_price` fn with an additional parameter pointing which price is being updated:
> 
> [aepp-sophia-examples/examples/ExchangeOracles/contracts/ExchangeMarket.aes](https://github.com/aeternity/aepp-sophia-examples/blob/0793a9f6e6f9b7b7d8e2fefc31aeda45f8ebb2d1/examples/ExchangeOracles/contracts/ExchangeMarket.aes#L46-L52)
> 
> Lines 46 to 52 in [0793a9f](/aeternity/aepp-sophia-examples/commit/0793a9f6e6f9b7b7d8e2fefc31aeda45f8ebb2d1)
> 
>  function update_ae_price(ae_price': int) = 
>    require(ae_price' > 0, "The price must be greater than zero") 
>    put(state{ae_price = ae_price'}) 
>   
>  function update_token_price(token_price': int) = 
>    require(token_price' > 0, "The price must be greater than zero") 
>    put(state{token_price = token_price'}) 
> Something like:
> 
> ```
> function update_price(new_price: int, asset: string) =
> ...
> put(state{assets[asset] = new_price})
> ```
> This would require some changes to the storage architecture as well e.g.
> 
> ```
> record state = { assets: map(string, int) }
> 
> ...
> 
> public stateful function init(ae_price: int, token_price: int) = 
>     require(ae_price > 0 , "The price for the AE must be greater than zero ")
>     require(token_price > 0 , "The price for the token must be greater than zero")
>     { assets = { ["ae"] = ae_price, ["token"] = token_price } }
> ```
> This will make the contract more generic with the ability to add more pricing related query options for other currencies if implemented such functionality.

Hey,
The idea of the contracts is that they are examples, and are used to show the functionalities of the oracles. The main idea behind the functions with predefined questions like "give me the ae price" is that the user doesn't have the ability to use this contracts for queries with a different purpose than the desired one.

I can agree with the other feedback that the update functions can be merged into one with one additional parameter, but I think that this will make the contracts a bit more complex. My goal was that the contracts are as simple as possible, but still representing the oracles functionalities.

**Comment by mradkov** [comment]
> > [aepp-sophia-examples/examples/ExchangeOracles/contracts/ExchangeMarket.aes](https://github.com/aeternity/aepp-sophia-examples/blob/0793a9f6e6f9b7b7d8e2fefc31aeda45f8ebb2d1/examples/ExchangeOracles/contracts/ExchangeMarket.aes#L20-L36)
> > Lines 20 to 36 in [0793a9f](/aeternity/aepp-sophia-examples/commit/0793a9f6e6f9b7b7d8e2fefc31aeda45f8ebb2d1)
> > function create_query(oracle: oracle_id, question: query_type, query_fee: int, query_ttl: int, relative_ttl: int) : query_id =
> > require(query_fee > 0 , "Query fee must be greater than zero")
> > require(query_ttl > 0 , "Query TTL must be greater than zero")
> > require(relative_ttl > 0 , "Relative TTL must be greater than zero")
> > Oracle.query(oracle, question, query_fee, RelativeTTL(query_ttl), RelativeTTL(relative_ttl))
> > function create_ae_price_query(oracle: oracle_id, query_fee: int, query_ttl: int, relative_ttl: int) : query_id =
> > require(query_fee > 0 , "Query fee must be greater than zero")
> > require(query_ttl > 0 , "Query TTL must be greater than zero")
> > require(relative_ttl > 0 , "Relative TTL must be greater than zero")
> > Oracle.query(oracle, "Give me the AE price in USD", query_fee, RelativeTTL(query_ttl), RelativeTTL(relative_ttl))
> > function create_token_price_query(oracle: oracle_id, query_fee: int, query_ttl: int, relative_ttl: int) : query_id =
> > require(query_fee > 0 , "Query fee must be greater than zero")
> > require(query_ttl > 0 , "Query TTL must be greater than zero")
> > require(relative_ttl > 0 , "Relative TTL must be greater than zero")
> > Oracle.query(oracle, "Give me the Token price in USD", query_fee, RelativeTTL(query_ttl), RelativeTTL(relative_ttl))
> > Why the need of 3 separate functions for the same purpose with the single difference of the question string, can't that be abstracted by using only the `create_query` fn?
> > Also, think this can be replaced with a single `update_price` fn with an additional parameter pointing which price is being updated:
> > [aepp-sophia-examples/examples/ExchangeOracles/contracts/ExchangeMarket.aes](https://github.com/aeternity/aepp-sophia-examples/blob/0793a9f6e6f9b7b7d8e2fefc31aeda45f8ebb2d1/examples/ExchangeOracles/contracts/ExchangeMarket.aes#L46-L52)
> > Lines 46 to 52 in [0793a9f](/aeternity/aepp-sophia-examples/commit/0793a9f6e6f9b7b7d8e2fefc31aeda45f8ebb2d1)
> > function update_ae_price(ae_price': int) =
> > require(ae_price' > 0, "The price must be greater than zero")
> > put(state{ae_price = ae_price'})
> > function update_token_price(token_price': int) =
> > require(token_price' > 0, "The price must be greater than zero")
> > put(state{token_price = token_price'})
> > Something like:
> > ```
> > function update_price(new_price: int, asset: string) =
> > ...
> > put(state{assets[asset] = new_price})
> > ```
> > This would require some changes to the storage architecture as well e.g.
> > ```
> > record state = { assets: map(string, int) }
> > 
> > ...
> > 
> > public stateful function init(ae_price: int, token_price: int) = 
> >     require(ae_price > 0 , "The price for the AE must be greater than zero ")
> >     require(token_price > 0 , "The price for the token must be greater than zero")
> >     { assets = { ["ae"] = ae_price, ["token"] = token_price } }
> > ```
> > This will make the contract more generic with the ability to add more pricing related query options for other currencies if implemented such functionality.
> 
> Hey,
> The idea of the contracts is that they are examples, and are used to show the functionalities of the oracles. The main idea behind the functions with predefined questions like "give me the ae price" is that the user doesn't have the ability to use this contracts for queries with a different purpose than the desired one.
> 
> I can agree with the other feedback that the update functions can be merged into one with one additional parameter, but I think that this will make the contracts a bit more complex. My goal was that the contracts are as simple as possible, but still representing the oracles functionalities.

I agree that the contracts should be as simple as possible, but shouldn't we make them as professional as possible as well. Since this is examples repo I think we can focus more on the best practices, since people will most likely copy/paste this and use the code however we've provided it.

If the idea of this repo is to become something like `open-zeppelin` I think we should be making more well-designed contracts - we can add comments in the code as much as we like and explain how exactly this is working, which I think will be useful for the user when trying the code.

Anyhow, this is my opinion only.

**Comment by thepiwo** [comment]
@maptuhec shall we merge?

**Comment by maptuhec** [comment]
> @maptuhec shall we merge?

if no one else wants to review it, I think we can merge

**Pull request merged**